cloudinit/distros/__init__.py

```diff
@@ -83,6 +83,30 @@ def __init__(self, name, cfg, paths):
         self.name = name
         self.networking = self.networking_cls()
 
+    def __setstate__(self, state):
+        """Restore instance state and handle missing attributes on upgrade.
+
+        This will be called when an instance of this class is unpickled; the
+        previous instance's `__dict__` is passed as `state`.  This method
+        restores that state into the current instance, and fixes up any
+        missing/incorrect state (usually caused by upgrades to cloud-init since
+        `state` was pickled).
+
+        See https://docs.python.org/3/library/pickle.html#object.__setstate__
+        for further background.
+        """
+        # Restore instance attributes
+        self.__dict__.update(state)
+        if "networking" not in self.__dict__ or not self.networking.__dict__:
+            # This is either a Distro pickle with no networking attribute OR
+            # this is a Distro pickle with a networking attribute but from
+            # before ``Networking`` had any state (meaning that
+            # Networking.__setstate__ will not be called).  In either case, we
+            # want to ensure that `self.networking` is freshly-instantiated:
+            # either because it isn't present at all, or because it will be
+            # missing expected instance state otherwise.
```

Comment:
In my reading around pickle/unpick and set-state;  since we're here, do we want to start versioning this in case we need to migrate to a different representation?

Reply:
It's not clear to me what global versioning would look like here: the pickle is generated recursively, so we don't have explicit control over what's included in it; I think any such version would be incremented a bunch, and as a result it would be very difficult to use it effectively.

For local/per-class versioning, we could implement `__getstate__` and have it set a static `"version": 1` in the state that gets stored in the pickle (which we would `.pop()` off before restoring the state in `__setstate__`). Then we would be able to express future state changes as `if version < 3: self.new_variable = "some value"` or similar. Is that what you had in mind?

One thing I would definitely like to figure out is how we can detect and review changes to the pickle representation that we generate. This may be something of a prerequisite to versioning anyway, because we need to know that/how our representation has changed to effectively version it. I wonder if something as simple as statically storing the "expected" pickle representation (i.e. the representation currently produced by cloud-init trunk) somewhere in tree and having a test which generates the pickle representation from the current codebase and fails if they don't match would do the trick. We'd (a) notice when a PR changed the representation, and (b) (assuming we store the `pickletools.dis` string representation rather than the binary one) have a reviewable diff of the change to the structure when people modified the "expected" pickle representation to get the test passing.

Reply:
I've put together a rough draft of what I described in that last para and pushed it to this PR (https://github.com/canonical/cloud-init/pull/609/files#diff-65e7350779a626d7cfabf2f3a54070cee4aaaa92a4e3f6ce3185b38416c68410).

Reply:
> For local/per-class versioning, we could implement __getstate__ and have it set a static "version": 1 in the state that gets stored in the pickle (which we would .pop() off before restoring the state in __setstate__). Then we would be able to express future state changes as if version < 3: self.new_variable = "some value" or similar. Is that what you had in mind?

The versioning of the pickled object makes sense to me. A couple of questions:
  1. when would we need to increment that version? Would you only increment if you had a different behavior in version 2 that was incompatible with the first upgrade path? Or would you increment with each de-serialization "fix" that needed applying?
 2.  Since the current class attribute definition on disk is what is loaded during un-serialize, would it make sense preserve that versioned dict of attrs and a tuple of (attr_test, (value or callable) )  to determine what what needs to be applied during  `__setstate__`.

```python
class Distro:
   cls_version = "2"
   # upgrade_attrs are applied at object de-serialization across upgrade path by __setstate__
   upgrade_attrs = {
       "1":
           {"networking": (
                _network_unset,  # test if true, apply callable below
                networking_cls   # set self.networking to result of callable if test above is True
             )
           }
    }
```

Reply:
> The versioning of the pickled object makes sense to me. A couple of questions:

Thanks for the review!

> 1. when would we need to increment that version? Would you only increment if you had a different behavior in version 2 that was incompatible with the first upgrade path? Or would you increment with each de-serialization "fix" that needed applying?

To some extent, I'm not sure. Introducing versioning now means that we have the option of using it in future, in the way that makes the most sense then. However, to be substantially less hand-wavy: I would anticipate that we would want to bump a class's version whenever new instance state is added to a class: that's generally what will prompt needing to apply a deserialization fix, so I think this may just be a different way of stating your final sentence.

(I could also imagine scenarios where we released a cloud-init with a faulty default, which we might also want to fixup on deserialisation: consider a hypothetical modification to the code to expect `blacklist_drivers` to be `[]` instead of `None` for the null case. We _could_ have all the code do `if self.blacklist_drivers in [None, []]` or whatever, _or_ we could replace `None` with `[]` during `__setstate__`. Whether or not that is the right thing to do is likely to vary case-by-case, and the effect on versioning is likely to be similarly specific.)

> 2. Since the current class attribute definition on disk is what is loaded during un-serialize, would it make sense preserve that versioned dict of attrs and a tuple of (attr_test, (value or callable) )  to determine what what needs to be applied during  `__setstate__`.
> 
> 
> ```python
> class Distro:
>    cls_version = "2"
>    # upgrade_attrs are applied at object de-serialization across upgrade path by __setstate__
>    upgrade_attrs = {
>        "1":
>            {"networking": (
>                 _network_unset,  # test if true, apply callable below
>                 networking_cls   # set self.networking to result of callable if test above is True
>              )
>            }
>     }
> ```

I appreciate the thought, but this feels like premature optimisation to me: we haven't done this enough to be sure that this is the right abstraction to apply (though I agree it looks like it fits what we know currently). If something along these lines is the right thing to do, then it should be fairly easy to migrate to it in future (once it will be reducing complexity); if it isn't easy to migrate to, then it probably isn't the right abstraction and we should discuss what _is_ at that point.